Add system-specific naming guidance (#1708)
open-telemetry/semantic-conventions: docs/general/naming.md

Metric `http.server.request.duration` uses attributes from the registry such as
`server.port`, `error.type`.

### System-specific attributes

**Status**: [Development][DocumentStatus]

When an attribute is specific to a particular system (e.g., project, provider, product),
the system name should be used in the attribute name, following the pattern:
`{optional domain}.{system}.*.{property}` pattern.

Examples:

> - `db.cassandra.consistency_level` - Describes the consistency level property

Author: Similar to https://github.com/open-telemetry/semantic-conventions/pull/1708/files#r1894523030

what's the benefit of having `db` in front of `cassandra` here? why not `cassandra.consistency_level`?

Author: Let's say we have some redis-specific attributes (and redis can be used as a DB and as a messaging system), would we define `db.redis` and `messaging.redis` ? probably not, it should be just `redis`.

Reviewer: Yeah in the particular case of `redis` it is clear that the prefix becomes kinda irrelevant.. but then again it is nice to have the well-known prefixes for the "main" semconv areas.. 🤔  I'm torn on this one.

Author: > but then again it is nice to have the well-known prefixes for the "main" semconv areas

could you elaborate? Is it nice to have `db.cassandra.*` ? Can you think of cases when it'd be useful?

Reviewer 2: From my perspective it's only 2 possible cases here:

1. the metric is cassandra specific so just the `cassandra.*` namespace is fine
2. the metric can be generalised across different database technologies and hence it's just `db.consistency_level`.

For databases, cloud providers, messaging systems this approach looks valid.
I'm not sure if it's totally equivalent for `system`, `process` and `linux` areas (see also https://github.com/open-telemetry/semantic-conventions/pull/1708#discussion_r1931790425).

Author: let's continue the discussion there. We should not drop `process` or `system` when they are meaningful, but we should do `linux` instead of `os.linux`

to the processes then to represent the count of the processes we can have a metric named
`system.process.count`.

#### Do not use `total`

UpDownCounters SHOULD NOT use `_total` because then they will look like
monotonic sums.

Counters SHOULD NOT append `_total` either because then their meaning will
be confusing in delta backends.

> ### Instrument naming

Reviewer 3: Why are we changing title convention everywhere?

Author: it's what otel.io does for link titles - https://github.com/open-telemetry/semantic-conventions/issues/169, pretty much all headings we have are in Sentence case now.

to the processes then to represent the count of the processes we can have a metric named
`system.process.count`.

#### Do not use `total`

UpDownCounters SHOULD NOT use `_total` because then they will look like
monotonic sums.

Counters SHOULD NOT append `_total` either because then their meaning will
be confusing in delta backends.

### Instrument naming

**Status**: [Development][DocumentStatus]

> **Status**: [Experimental][DocumentStatus]

Author: great catch, thanks!

   - Independent projects that do not belong to a specific company, such as Apache Foundation projects like
     `kafka` or `cassandra`.
   - Products with names similar to the owning company, such as `mongodb` or `elasticsearch`
   - Widely recognized products that are popular outside their company's ecosystem.
     These products often have trademarks without the company name and have
     their own top-level domain (e.g. `spring` or `mysql`).

4. In other cases, the system name SHOULD be prefixed with the company (organization,
   division, or group) name. For cloud services, the name SHOULD use the
   corresponding cloud provider name. For example, use `aws.dynamodb` or `azure.cosmosdb`.

   The company (organization, division, or group) name SHOULD remain consistent
   across multiple product names in different semantic convention areas.

### System-specific Attributes

When an attribute is specific to a particular system (project, provider, product),
the corresponding attribute name SHOULD start with the system name following the
`{system_name}.*.{property}` pattern.

Examples:

> - `cassandra.consistency.level` - Describes the consistency level property

Reviewer 2: It'd be nice to explain here that if a metric/attribute can be generalized across different implementations/technologies then `db.consistency.level` could be used.

In general if we could collect more examples here to illustrate the different areas/use-cases that would be great.

Author: absolutely, this is what I'm trying to do in https://github.com/open-telemetry/semantic-conventions/pull/1707 and would be interested in your input

## System-specific naming

**Status**: [Development][DocumentStatus]

### System (project/product/provider) name attribute

Semantic conventions for a certain area are usually applicable to multiple systems
(projects, providers, products).

For example, database semantic conventions can be used to describe telemetry for a
broad range of database systems.

Such conventions SHOULD define an attribute to represent system name following
`{area}.system|provider|protocol.name` pattern.

For example, database conventions include the `db.system.name` attribute.
<!-- not yet, we should not merge it before https://github.com/open-telemetry/semantic-conventions/pull/1613 -->

### Choosing a System Name

When adding new system to the semantic conventions, follow these principles in descending order of priority:

1. The system name SHOULD adhere to the general attribute naming guidelines outlined in this document,
   as it will be used as a namespace in [system-specific attribute names](#system-specific-attributes).

2. The system name SHOULD unambiguously identify this specific product or project.

   For example, use `gcp.pubsub` or `oracle.db`. Avoid generic names like `pubsub`,

>    which could refer to multiple messaging products, or `oracle` which lacks specificity.

Reviewer 4: ```suggestion
   which could refer to multiple messaging products, or `oracle` which could refer to
   multiple Oracle products.
```

3. The system name SHOULD match the corresponding project or product name in the following cases:

   - Independent projects that do not belong to a specific company, such as Apache Foundation projects like
     `kafka` or `cassandra`.
   - Products with names similar to the owning company, such as `mongodb` or `elasticsearch`
   - Widely recognized products that are popular outside their company's ecosystem.
     These products often have trademarks without the company name and have
     their own top-level domain (e.g. `spring` or `mysql`).

3. In other cases, the system name SHOULD be prefixed with the company (organization,
   division, or group) name. For cloud services, the name SHOULD use the

>    corresponding cloud provider name. For example, use `ibm.db2` or `aws.dynamodb`.

Reviewer 4: `ibm.db2` might not be the best example since some (me) don't think of db2 primarily as a cloud service (i'm probably wrong at this point...), and it's not clear to me whether ibm cloud will have its own namespace (`ibm.cloud`?)

```suggestion
   corresponding cloud provider name. For example, use `aws.dynamodb` or `azure.cosmosdb`.
```

   across multiple product names in different semantic convention areas.

### System-specific Attributes

When an attribute is specific to a particular system (project, provider, product),
the corresponding attribute name SHOULD start with the system name following the
`{system_name}.*.{property}` pattern.

Examples:

- `cassandra.consistency.level` - Describes the consistency level property
  specific to the Cassandra database.
- `aws.s3.key` - Refers to the `key` property of the AWS S3 product.
- `signalr.connection.status` – Indicates the connection status of the SignalR
  network protocol.

The value of the [`*.system.name`](#attributes-that-describe-system-name) (or similar)
attribute MUST match the root namespace used in the system specific attribute being defined.

For example, both `cassandra.consistency.level` attribute name and `db.system.name=cassandra`
 use the same system name (`cassandra`).

### System-specific Metrics

When a metric is specific to a system (project, provider, product) in a given area,
the system name should be included in the instrument name following the
`{system_name}.*.{metric_name}` pattern.

> For example, `azure.cosmosdb.client.operation.request_charge`

Author: ```suggestion
For example, `azure.cosmosdb.client.operation.request.charge`
```

<!-- todo not yet - https://github.com/open-telemetry/semantic-conventions/pull/1711 -->
<!-- update when https://github.com/open-telemetry/semantic-conventions/pull/1613 is merged -->

The value of the [`*.system.name`](#attributes-that-describe-system-name) (or similar)
attribute MUST match system specific metric namespace.

> For example, both `azure.cosmosdb.client.operation.request_charge` metric and `db.system.name=azure.cosmosdb`

Reviewer 4: ```suggestion
For example, both the `azure.cosmosdb.client.operation.request_charge` metric and the `db.system.name=azure.cosmosdb`
```

Units do not need to be specified in the names since they are included during
instrument creation, but can be added if there is ambiguity.

### Client and server metrics

Metrics that measure aspects of a physical or logical network call should include
an indication of the side being measured when ambiguity exists.

In such cases, the metric name should follow the pattern: `{domain}.{client|server}.{metric_name}`.

For example:
- `http.client.request.duration`
- `gen_ai.server.request.duration`
- `messaging.client.sent.messages`
- `messaging.process.duration` - the term `process` clearly indicates that
  the metric is reported by the consumer.
- `kestrel.connection.duration` - here, `kestrel` is the name of the web server,
  so no additional indication is necessary.

### System-specific metrics

When a metric is specific to a system (e.g., project, provider, product) in a given domain,
the system name should be included in the instrument name using the pattern:
`{domain}.{client|server}.{system}.*.{property}` pattern.

> For example, `db.client.cosmosdb.operation.request_charge`

Author: the difference with `signalr.server.connection.duration` below makes me think if we should simplify it and do `cosmosdb.client.operation.request_charge`. What's the benefit of having `db` in front of it?

Author: summoning @open-telemetry/semconv-system-approvers 

It's the same question as for `system.linux` vs `linux` metrics in https://github.com/open-telemetry/semantic-conventions/issues/1403 or https://github.com/open-telemetry/semantic-conventions/pull/1618

Reviewer 2: The difference with `process.linux.cgroup` is that the name describes effectively a cgroup that belongs to a linux process. While `cosmosdb.client.operation.request_charge` can stand on its own without the need of `db` in front.

Maybe we can try and standardize this difference as part of this doc?

Reviewer 4: what about `linux.process.cgroup`?

Author: Examples:

- [Imaginary] `linux.distro.name` (`linux.distro.version`, `linux.kernel.version`) instead of `os.linux.distro.name`
- [Real] [`jvm.system.cpu.load_1m`](https://github.com/open-telemetry/semantic-conventions/blob/a72c5eea291b947bcd4916b70feb3b2e6b4eae9a/model/jvm/metrics-experimental.yaml#L24) instead of `system.jvm..cpu.load_1m``
- [Real] [`dotnet.process.cpu.count`](https://github.com/open-telemetry/semantic-conventions/blob/a72c5eea291b947bcd4916b70feb3b2e6b4eae9a/model/dotnet/runtime-metrics.yaml#L4) instead of `process.dotnet.cpu.count`
 -  [Real, but different pattern] `linux.memory.slab.usage` (or `linux.system.memory.slab.usage`) instead of [`system.linux.memory.slab.usage`](https://github.com/open-telemetry/semantic-conventions/blob/a72c5eea291b947bcd4916b70feb3b2e6b4eae9a/model/system/metrics.yaml#L396)

Justification for `linux.process.cgroup` - cgroup is a property of a process, not linux. E.g. we have `process.pid`. What if there was a linux-specific process id, would we do `linux.process.id` or `process.linux.id` ? I think former is more friendly.

And since `linux`-specific metrics fall into 'specialist case' bucket, they don't necessarily need to be discoverable when you type `process` and let autocomplete guide you

Author: I added `Known exceptions` section and mentioned system.{os} and process.{os} there. As discussed with @braydonk during semconv call, we should follow-up with the PR explaining the reasoning for it.
